plugins/org.eclipse.elk.core/src/org/eclipse/elk/core/util/nodespacing/internal/algorithm/PortPlacementCalculator.java

```diff
@@ -0,0 +1,369 @@
+/*******************************************************************************
+ * Copyright (c) 2017 Kiel University and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *    Kiel University - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.elk.core.util.nodespacing.internal.algorithm;
+
+import org.eclipse.elk.core.math.ElkPadding;
+import org.eclipse.elk.core.math.ElkRectangle;
+import org.eclipse.elk.core.math.KVector;
+import org.eclipse.elk.core.options.CoreOptions;
+import org.eclipse.elk.core.options.PortAlignment;
+import org.eclipse.elk.core.options.PortConstraints;
+import org.eclipse.elk.core.options.PortSide;
+import org.eclipse.elk.core.options.SizeOptions;
+import org.eclipse.elk.core.util.adapters.GraphAdapters.PortAdapter;
+import org.eclipse.elk.core.util.nodespacing.internal.NodeContext;
+import org.eclipse.elk.core.util.nodespacing.internal.PortContext;
+import org.eclipse.elk.core.util.nodespacing.internal.cellsystem.AtomicCell;
+import org.eclipse.elk.graph.properties.IProperty;
+import org.eclipse.elk.graph.properties.Property;
+
+/**
+ * Actually places ports.
+ */
+public final class PortPlacementCalculator {
+    
+    /**
+     * Copy of the {@link de.cau.cs.kieler.klay.layered.properties.InternalProperties#PORT_RATIO_OR_POSITION}
+     * option. For further information see the documentation found there. We added this copy here to
+     * allow a generic treatment of spacing calculations for graph elements. See the
+     * {@link org.eclipse.elk.core.util.nodespacing} package. [programmatically set]
+     * 
+     * TODO This should be replaced by a proper property declared in a MELK file.
```

Comment:
Not right away, this is a bit more complicated. This constant was an implementation detail of ELK Layered. Setting the ratio directly with a layout option was originally not the idea, but could be added. Should then be supported by ELK Layered as an official way to set a port's aspect ratio as well, though.